virt: more network support

cbosdo opened the pull request:
### What does this PR do?

Add more parameters for virtual network creation in `virt.network_defined` and `virt.network_running` states.

### What issues does this PR fix or reference?
Fixes: #59143

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [X] Docs
- [X] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

cbosdo committed:
Fix indentation of Jinja instructions in libvirt_domain.jinja

cbosdo committed:
Use more opt_attribute macro in libvirt_domain.jinja

cbosdo committed:
Unify XML indentation in libvirt_domain.jinja

cbosdo committed:
Move virt network generation tests to pytest

cbosdo committed:
Extract XML space stripping function from virt module

Stripping spaces and indentation from XML could also be useful in other
places, moving to xmlutil to help reuse.

cbosdo committed:
Fix link in virt state documentation

cbosdo committed:
Extract the XML cleanup code from virt.pool_update

The network_update code will be rather similar to the pool_update one.
In order to share the XML tree cleanup for easier comparisons, create a
helper function in the virt module.

cbosdo committed:
virt: expose more properties in virt.network_define

In order to let users define more types of virtual networks, expose more
of the libvirt virtual network properties.

cbosdo committed:
Remove useless code in virt pytest fixture

cbosdo committed:
Add virt.network_update function

In order to enhance the virt.network_defined state, a function to test
if an update is needed and update the network is needed. This is done by
the newly added virt.network_update function.

cbosdo committed:
Convert the virt network state unit tests to pytest

Converting these tests helped reducing the number of lines of code
thanks to the pytest parametrize feature. This is also the occasion to
split the big tests into smaller ones to report more meaningfull errors
and make it more readable.

cbosdo committed:
Let virt.network_update state change existing networks

Instead of simply reporting existing networks, update them if needed
like other states. Also bubble up the new properties from the
virt.define() function.

cbosdo committed:
Fix the new iothreads virtual disk parameter

io_uring has recently been added as another IO policy on virtual disks.
Keep the parameter opened for changes. Also add the optional iothread ID
in case the user wants to pin a disk to some IO thread (and thus to a
CPU).

cbosdo committed:
Add virt.node_devices function

For the user to be able to pass host devices through he needs to get a
list of the devices that can be passed.

cbosdo committed:
virt: add PCI and USB host devices support to virt init and update

In quite a few cases it may be useful to pass a PCI or USB device from
the host to the VM. Add support for this in the virt.init() and
virt.update() functions.

cbosdo committed:
Convert virt domain state unit tests to pytest

While converting the virt domain-related states to pytest I realized
the __opts__["test"] == False case was not handled in some of them.
This commit also fixes the return code for virt.shutdown,
virt.powered_off, virt.snapshot and virt.rebooted states. It also
prevents the actual call to be issued in test mode.

cbosdo committed:
Add host_devices to virt running and defined states

Expose the new host_devices parameter to the virt.running and
virt.defined states.

cbosdo committed:
Remove deprecated update parameter in virt.defined and virt.running

cbosdo commented:
This PR is based on #59144

cbosdo commented:
re-run pr-macosxcatalina-py3-pytest

cbosdo committed:
Convert virt _diff_nics() unit test to pytest

cbosdo committed:
virt: better compare NICs of running VMs

On a running guest, libvirt changes the XML definition of the network
interfaces of type "network" to the type of the network (for instance
bridge). In such a case the virt.update() function will find the two
NICs different even if they may not be... so we need to try harder to
compare.

cbosdo committed:
virt: hostdev network fixes

A network with hostdev forward mode has no bridge and no mac. So we need
to handle this in a few places in the virt module.

cbosdo committed:
virt: extract the live update code from the update function

In order to help reusing the device changes computing code and avoid
getting a giant virt.update(), move the live update code of it into a
specific internal function.

cbosdo committed:
virt: better handle comparison of hostdev NIC interfaces

When a domain has a NIC of type network pointing to a network with hostdev
forward, libvirt changes its running XML definition with a hostdev
interface with a PCI address from those in the network.

Handle this case to avoid useless interface detaching / attaching.

waynew commented:
Well, this is quite the monster PR :upside_down_face: 

I'm OOO through the end of the year - I don't know how antsy you are to get this in, but if nobody reviews before I get back I'll plan to take a look at this Jan 4

cbosdo commented:
I'ld like to have it in Aluminium ideally, so can still wait for you to come back I guess

cbosdo committed:
virt: better compare hostdev networks

Libvirt adds the PCI addresses of the SR-IOV device virtual functions
when only providing the physical function. Those need to be removed in
order to avoid network changes for no reason in virt.network_update()

waynew commented on salt/utils/xmlutil.py:
```diff
@@ -379,3 +379,35 @@ def change_xml(doc, data, mapping):
+    :return: the node
+    """
+
+    def space_stripper(node):
+        if node.tail is not None:
+            node.tail = node.tail.strip(" \t\n")
+        if node.text is not None:
+            node.text = node.text.strip(" \t\n")
```
**suggestion**: I don't think that we gain anything by having this function as an internal function. We could either make this an internal function, e.g.

```
def _space_stripper(...):
```

Or, if we're going to use recursion we can just simplify this:

```
def strip_spaces(node):
    '''...'''

    if node.tail is not None:
        node.tail = node.tail.strip(" \t\n")
    if node.text is not None:
        node.text = node.text.strip(" \t\n")
    try:
        for child in node:
             strip_spaces(child)
    except RecursionError:
        raise Exception('but something better, OK?')

    return node
```

by removing the `visit` call altogether.

cbosdo replied:
OK, I'll go for the version without the `visit` call: looks cleaner

waynew commented on salt/modules/virt.py:
```diff
@@ -6796,19 +7340,287 @@ def network_define(name, bridge, forward, ipv4_config=None, ipv6_config=None, **
+            if not child.tail and not child.text and not child.items() and not child:
+                node.remove(child)
+
+    return xmlutil.visit(node, empty_node_remover)
```
**suggestion**: something feels off about creating an internal function to allow recursion. This function could also directly recurse:

```
def _remove_empty_xml_node(node):
    """
    Remove the nodes with no children, no text and no attribute
    """
     for child in node:
         if not child.tail and not child.text and not child.items() and not child:
             node.remove(child)
         else:
             _remove_empty_xml_node(child)

    return node
   ```

cbosdo replied:
:+1:

waynew commented on salt/modules/virt.py:
```diff
@@ -1179,23 +1214,73 @@ def _gen_vol_xml(
     """
     Generate the XML string to define a libvirt network
     """
+    vport_context = vport
+    if isinstance(vport, str):
+        vport_context = {"type": vport}
```
*suggestion*: An if/else is clearer here, since `vport_context` is falling back to `vport`.
```suggestion
    if isinstance(vport, str):
        vport_context = {"type": vport}
    else:
        vport_context = vport
```

waynew replied:
Definitely better than `vport_context = {"type": vport} if isinstance(vport, str) else vport` :nauseated_face:

cbosdo replied:
OK, I'll change to if/else

waynew commented on salt/modules/virt.py:
```diff
@@ -1179,23 +1214,73 @@ def _gen_vol_xml(
+        vport_context = {"type": vport}
+
+    tag_context = tag
+    if isinstance(tag, str) or isinstance(tag, int):
```
**suggestion** isinstance can take a tuple:

```
if isinstance(tag, (str,int)):
    tag = {"tags": [...]}
```

cbosdo replied:
thanks for the tip, I wasn't aware of that!

waynew commented on salt/modules/virt.py:
```diff
@@ -6744,17 +7148,152 @@ def network_define(name, bridge, forward, ipv4_config=None, ipv6_config=None, **
+                - addr: 8.8.8.8
+                - domain: www.example.com
+              txt:
+                host.widgets.com.: "printer=lpr5"
```
**question**: Should this also be something like `widgets.example.com`?

cbosdo replied:
this is purely left to dnsmasq... so I guess that can also be it

waynew replied:
I only raise the issue because I read somewhere recently that example documentation should be examples :upside_down_face: 

Basically -- putting "real" things in increases the chances that someone accidentally does a copy/paste and horks something up. Keeping things as example domains, [IP addresses](https://tools.ietf.org/html/rfc5737), etc, helps to avoid that problem. If it were the only thing on a PR I probably wouldn't bring it up, but I was already in here thinking :sweat_smile:

cbosdo replied:
I'm no DNS expert... so I'll put there the values you think are best. IIRC I picked that one from some libvirt example file

cbosdo replied:
I'll add the examples from https://github.com/imp/dnsmasq/blob/master/dnsmasq.conf.example to be on the safe side

waynew reviewed (changes requested):
I'm about 25% through my review on this, but I'm off for the day -- I'll pick this up tomorrow AM Central Time, most likely.

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2622,9 +2715,17 @@ def init(
                       hostname_property: virt:hostname
                       sparse_volume: True
 
-    iothreads
-        When ``True`` dedicated threads will be used for the I/O of the disk.
-        (Default: ``False``)
```
no, there is no need since it hasn't been released yet. And this change is already part of ACKed PR #59144

waynew reviewed (changes requested):
**praise**: I have to say overall, *great job*! This was clearly a lot of coding and testing :raised_hands: 

Most of the feedback I have are just nitpicks - I didn't see any logic failures, and only one spot that had me :suspect: 

Once you've addressed the feedback (either :+1: the suggestions or :wastebasket:), I'm :+1: on this. Thank you so much, and again, great job!

cbosdo commented on salt/modules/virt.py:
```diff
@@ -6796,19 +7340,287 @@ def network_define(name, bridge, forward, ipv4_config=None, ipv6_config=None, **
+        a possible DHCP configuration. The structure is documented
+        in net-define-ip_.
+
+    :type ipv6_config: dict or None
+
+    :param connection: libvirt connection URI, overriding defaults
+    :param username: username to connect with, overriding defaults
+    :param password: password to connect with, overriding defaults
```
added dots at the end of the sentences in the added documentation. Won't change the existing now though

cbosdo commented on tests/pytests/unit/states/virt/test_domain.py:
```diff
@@ -0,0 +1,836 @@
+                "comment": "Domain myvm unchanged",
+            } == virt.defined("myvm")
+            init_mock.assert_not_called()
+            assert 1 == len(calls_with_params(update_mock, test=test))
```
That would be possible... and we had that before. The issue is that every time a parameter is added we need to update those tests to add it too. In a sense it's not a bad thing... but it's adding lots of line to those tests.

cbosdo replied:
or maybe I should write a wrapper to create the calls and add the missing parameters set to `None`... would be a nice trade off

cbosdo replied:
done with the wrapper helper function: much cleaner :)

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1136,6 +1139,38 @@ def _gen_xml(
+    # Process host devices passthrough
+    hostdev_context = []
+    for hostdev_name in host_devices or []:
+        hostdevice = conn.nodeDeviceLookupByName(hostdev_name)
```
yes it can error out. One of its uses is in a try/except block catching libvirt errors... which is fine. The other doesn't. I'll add a try/except at this level too then.

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2956,6 +3068,31 @@ def _filter_graphics(gfx):
+        When the domain is running, the hostdevs element may contain additional properties.
+        This function will only keep the ones we care about
+        """
+        definition = {
+            "type": dev.get("type"),
+        }
+        if definition["type"] == "pci":
+            address_node = dev.find("./source/address")
+            for attr in ["domain", "bus", "slot", "function"]:
+                definition[attr] = address_node.get(attr)
+        elif definition["type"] == "usb":
+            for attr in ["vendor", "product"]:
+                definition[attr] = dev.find("./source/" + attr).get("id")
+        return definition
```
I went with `type_` looks better with the code editor syntax highlighting and who knows if that could be ruled out by pylint.

cbosdo committed:
Add xmlutil function dumping a node into a string

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3151,6 +3298,224 @@ def _diff_console_list(old, new):
+                "device": "disk",
+                "cmd": "updateDeviceFlags",
+                "args": [
+                    salt.utils.stringutils.to_str(ElementTree.tostring(updated_disk))
```
:+1:

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3151,6 +3298,224 @@ def _diff_console_list(old, new):
+                    "detachDevice": "detached",
+                    "updateDeviceFlags": "updated",
+                }
+                if device_type not in status:
+                    status[device_type] = {}
```
another function I wasn't aware of! Thanks for the hint, makes it cleaner :+1:

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3151,6 +3298,224 @@ def _diff_console_list(old, new):
+
+    for cmd in commands:
+        try:
+            ret = getattr(domain, cmd["cmd"])(*cmd["args"]) if not test else 0
```
:+1:

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3151,6 +3298,224 @@ def _diff_console_list(old, new):
+            ret = getattr(domain, cmd["cmd"])(*cmd["args"]) if not test else 0
+            device_type = cmd["device"]
+            if device_type in ["cpu", "mem"]:
+                status[device_type] = not bool(ret)
```
:+1:

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3806,26 +4175,24 @@ def _set_cpuset(node, value):
+    func_locals = locals()
+
+    def _skip_update(names):
+        return not bool([n for n in names if func_locals.get(n) is not None])
```
the whole statement can even be changed to `return all(func_locals.get(n) is None for n in names)`. Thanks for the reminder: I keep forgetting `any()` and `all()`

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3858,145 +4225,15 @@ def _set_cpuset(node, value):
+        if errors:
+            if "errors" not in status:
+                status["errors"] = []
+            status["errors"] += errors
```
Addressed all the comments. Thanks for the thorough review!

cbosdo commented:
Just fixed the rebase on latest master to get the new test pass again.

cbosdo commented:
re-run pr-amazon2-py3-pytest

cbosdo committed:
Merge branch 'master' into virt-network

cbosdo committed:
Merge branch 'master' into virt-network

cbosdo commented:
re-run freebsd122-py3-pytest

waynew reviewed (approved):
:rocket:

cbosdo committed:
Merge branch 'master' into virt-network

cbosdo commented:
re-run freebsd122-py3-pytest

cbosdo commented:
re-run ubuntu2004-py3-pytest

cbosdo committed:
Merge branch 'master' into virt-network

cbosdo commented:
@dwoz could you merge this PR while it's Green and ACKed?

cbosdo committed:
Merge branch 'master' into virt-network

cbosdo commented:
re-run freebsd122-py3-pytest
